cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -0,0 +1,179 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/flight/sql/server_session_middleware.h"
+#include <boost/lexical_cast.hpp>
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
```

Comment:
I don't believe Boost is always available. If you want to use it, we'll have to adjust the CMakeLists to make sure it's found, and adjust ThirdpartyToolchain.cmake to reflect the new dependency.

If there's a small/vendorable UUID implementation available, that might be preferable. CC @pitrou

Reply:
There are many UUID libraries available, but there are a limited number that work as well as boost. Most do not randomize with 128-bits of precision (see this discussion: https://stackoverflow.com/questions/24365331/how-can-i-generate-uuid-in-c-without-using-boost-library). Others have a C++17 requirement -- would that be OK to use in the Arrow project?

[crossguid](https://github.com/graeme-hill/crossguid) utilizes operating system calls for generating UUIDs so it should generate them securely enough. It supports Windows, Mac, Linux, and Android -- is this enough platform support for Arrow?

Reply:
Why do we need UUIDs at all?

Reply:
C++17 is ok.

Antoine is right, strictly speaking, we could make this implementation take a generator for unique IDs, and a server implementation could plug a UUID (or other) generator into it. For unit tests, we can just use a random string. Let's do that @indigophox and avoid the dependency entirely.

Reply:
Absolutely agreed, on it.

Reply:
Done.